### PR TITLE
Add CUDA CFG support

### PR DESCRIPTION
**Pull request by Matt711**
I'm building some jupyter magic tooling as described in #8960. I need to be able to display the control flow graph for CUDA kernels. This PR adds an `inspect_cfg` method to CUDA kernels that returns the CFG as text in the dot language form. I plan on piping this output to the dot command line tool to display the digraph as an image in Jupyter notebooks. Ex:

```python
from numba import cuda
from numba.types import int32
import numpy as np

sig = (int32[::1], int32[::1])

@cuda.jit(sig)
def add(x, out):
    i = cuda.grid(1)
    if i < len(x):
        out[i] += x[i]
    
a = np.arange(8, dtype=np.float32)
d_a = cuda.to_device(a)
d_out = cuda.device_array_like(d_a)

add[2,4](d_a, d_out)
print(next(iter(add.inspect_cfg().items()))[1])
```

```dot
digraph f {
subgraph "add" {
node [fontname="Courier",fontsize=10,shape=Mrecord];
"add"
[add:\l\ \ MOV\ R1,\ c\[0x0\]\[0x28\]\ ;\l\ \ S2R\ R0,\ SR_TID.X\ ;\l\ \ S2R\ R3,\ SR_CTAID.X\ ;\l\ \ IMAD\ R0,\ R3,\ c\[0x0\]\[0x0\],\ R0\ ;\l\ \ ISETP.GE.U32.AND\ P0,\ PT,\ R0,\ c\[0x0\]\[0x188\],\ PT\ ;\l\ \ SHF.R.S32.HI\ R2,\ RZ,\ 0x1f,\ R0\ ;\l\ \ ISETP.GE.AND.EX\ P0,\ PT,\ R2,\ c\[0x0\]\[0x18c\],\ PT,\ P0\ ;\l|<exit0>\ \ @P0\ EXIT\ ;\l|<exitpost0>\ \ ISETP.GE.AND\ P0,\ PT,\ R0,\ RZ,\ PT\ ;\l\ \ ULDC.64\ UR4,\ c\[0x0\]\[0x118\]\ ;\l\ \ SEL\ R5,\ RZ,\ c\[0x0\]\[0x1c0\],\ P0\ ;\l\ \ SEL\ R3,\ RZ,\ c\[0x0\]\[0x188\],\ P0\ ;\l\ \ IADD3\ R6,\ P1,\ R0.reuse,\ R5,\ RZ\ ;\l\ \ IADD3\ R0,\ P2,\ R0,\ R3,\ RZ\ ;\l\ \ SEL\ R7,\ RZ,\ c\[0x0\]\[0x1c4\],\ P0\ ;\l\ \ SEL\ R3,\ RZ,\ c\[0x0\]\[0x18c\],\ P0\ ;\l\ \ IMAD.X\ R7,\ R2,\ 0x1,\ R7,\ P1\ ;\l\ \ LEA\ R4,\ P1,\ R0,\ c\[0x0\]\[0x180\],\ 0x2\ ;\l\ \ IMAD.X\ R3,\ R2,\ 0x1,\ R3,\ P2\ ;\l\ \ LEA\ R2,\ P0,\ R6,\ c\[0x0\]\[0x1b8\],\ 0x2\ ;\l\ \ LEA.HI.X\ R5,\ R0,\ c\[0x0\]\[0x184\],\ R3,\ 0x2,\ P1\ ;\l\ \ LEA.HI.X\ R3,\ R6,\ c\[0x0\]\[0x1bc\],\ R7,\ 0x2,\ P0\ ;\l\ \ LDG.E\ R4,\ \[R4.64\]\ ;\l\ \ LDG.E\ R7,\ \[R2.64\]\ ;\l\ \ IMAD.IADD\ R7,\ R4,\ 0x1,\ R7\ ;\l\ \ STG.E\ \[R2.64\],\ R7\ ;\l|<exit1>\ \ EXIT\ ;\l}"]
}
}
```

### CODE REVIEW

**Comment by gmarkall** [comment]
gpuci run tests

File: numba/cuda/dispatcher.py [hunk 1]
```diff
@@ -979,6 +987,28 @@ def inspect_asm(self, signature=None):
                 return {sig: overload.inspect_asm(cc)
                         for sig, overload in self.overloads.items()}
 
+    def inspect_cfg(self, signature=None):
```
Comment (gmarkall) [inline numba/cuda/dispatcher.py:990]: Dispatcher objects already have an `inspect_cfg()` method, which is supposed to return the CFG: https://numba.readthedocs.io/en/stable/reference/jit-compilation.html#Dispatcher.inspect_cfg - this method will need another name as it presently shadows it.

It does happen that `inspect_cfg()` is broken for CUDA dispatchers, but that's a separate issue.

**Review by gmarkall — CHANGES_REQUESTED** [review]
Thanks for the PR! In general the idea is good - there are a few comments on the diff.

File: numba/cuda/tests/cudapy/test_inspect.py [hunk 2]
```diff
@@ -148,6 +148,21 @@ def f(x):
 
         self.assertIn('nvdisasm is required', str(raises.exception))
 
+    @skip_without_nvdisasm('nvdisasm needed for inspect_cfg()')
+    def test_inspect_cfg(self):
```
Comment (gmarkall) [inline numba/cuda/tests/cudapy/test_inspect.py:152]: For consistency with the name of the method under test:

```suggestion
    @skip_without_nvdisasm('nvdisasm needed for inspect_sass_cfg()')
    def test_inspect_sass_cfg(self):
```

**Comment by gmarkall** [comment]
gpuci run tests

File: docs/upcoming_changes/9051.cuda.rst [hunk 3]
```diff
@@ -0,0 +1,8 @@
+
+Add CUDA SASS CFG Support
+=========================
+
+This PR This PR adds support for getting the SASS CFG in dot language format. 
+It adds an inspect_sass_cfg method to CUDADispatcher and the -cfg flag to
+the nvdisasm command line tool.
```
Comment (gmarkall) [inline docs/upcoming_changes/9051.cuda.rst:7]: For rst I think we can highlight these as code (and add brackets to the function name) with:

```suggestion
It adds an ``inspect_sass_cfg()`` method to CUDADispatcher and the ``-cfg``
flag to the nvdisasm command line tool.
```

**Comment by gmarkall** [comment]
gpuci run tests

**Review by gmarkall — APPROVED** [review]
Many thanks for the PR!

@esc Could this have a CUDA buildfarm run please?

**Comment by esc** [comment]
`Build numba_smoketest_cuda_yaml_208 has started`

**Comment by esc** [comment]
I am getting the following error:

```
======================================================================
FAIL: test_inspect_sass_nvdisasm_missing (numba.cuda.tests.cudapy.test_inspect.TestInspect)
----------------------------------------------------------------------
Traceback (most recent call last):
File "/opt/conda/envs/testenv_a0a80147/lib/python3.9/site-packages/numba/cuda/tests/cudapy/test_inspect.py", line 149, in test_inspect_sass_nvdisasm_missing
self.assertIn('nvdisasm is required', str(raises.exception))
AssertionError: 'nvdisasm is required' not found in 'nvdisasm has not been found. You may needto install the CUDA toolkit and ensure thatit is available on your PATH.\n'

```

**Comment by gmarkall** [comment]
@esc Many thanks - the test case needs updating, this is a genuine problem with the PR.

**Review by gmarkall — CHANGES_REQUESTED** [review]
@Matt711 Can you update the expected message in `test_inspect_sass_nvdisasm_missing` please?

**Comment by stuartarchibald** [comment]
@gmarkall the dispatcher base object has `inspect_disasm_cfg()` which I think shows something similar/equivalent to the SASS CFG developed here (`inspect_disasm_cfg` shows a CFG of the disassembled object code). Should these pieces of functionality share the same method name or is it better to keep them separate as they are at present e.g. `inspect_sass_cfg()` is CUDA specific, `inspect_disasm_cfg` is CPU? If they are kept separate then perhaps longer term the `inspect_disasm_cfg` functionality really ought to be made a method on the CPU dispatcher class and not the dispatcher base class.

**Comment by gmarkall** [comment]
@stuartarchibald Because `inspect_asm()` shows the PTX, I'd expect `inspect_disasm_cfg()` to give a CFG for the PTX. `inspect_sass()` gives the SASS for a kernel, so I think `inspect_sass_cfg()` is the right name here. If we find a way to generate a PTX CFG in future then I'd suggest that should be called `inspect_disasm_cfg()`.

**Review by gmarkall — APPROVED** [review]
Assuming this is OK on the buildfarm this time, this still looks good to me.

**Comment by gmarkall** [comment]
@esc Could this have another buildfarm run please?

**Comment by esc** [comment]
`Build numba_smoketest_cuda_yaml_209 has started`

**Comment by sklam** [comment]
> `Build numba_smoketest_cuda_yaml_209 has started`

passed

**Pull request merged**